### PR TITLE
#258

### PR DESCRIPTION
**Pull request by TheManuz**
I've finally been able to modify the event system.
##### pointer.js
I needed to modify the structure that keeps the reference of handlers for many reasons:
* when a _pointermove_ event goes out or inside the bounds of a rect, when need to fire _pointerleave_ or _pointerenter_ events. But for this to happen we need to process _pointermove_ events even if the object is not listening for them, because it may be listening for _pointerleave_ or _pointerenter_ instead.
* we needed a way to access to all the callbacks tied to a specific __rect__, in order to dispatch the right event
* we needed every callback to being able to access the pointerId, in order to know if the object is already clicked or not and by what touch

The "evtHandlers" structure is now a dictionary where the objects gets saved by their GUID (they are assigned one if they don't have it). For each object we save:
```javascript
evtHandlers[rect.GUID] = {
	rect : rect,		// the rect reference
	callbacks : {},		// a callbacks dictionary (to be accessed with the event names)
	pointerId : null,	// needed at this level, accessible for every callback
};
```
The callback dictionary contains, for each event name, an array containing callbacks.
Every time a pointerEventCallback gets registered, it's pushed into this array (this is pretty much how it was before, just a level deeper into the structure):
```javascript
evtHandlers[rect.GUID].callbacks[eventType].push(callback);
};
```
##### GUI.js
I also modified the GUI_Object to use the new event system.
I tested it on Google Chrome and on CocoonJS on Android (it finally handles correctly when a touch slides off the screen: it fires a "pointerCancel" and the button goes unpressed. It also behaved good when sliding a finger between 2 buttons: fires "pointerEnter" and "pointerLeave", the buttons get pressed and unpressed correctly).

It may need other tests, though.

### CODE REVIEW

**Comment by parasyte** [comment]
@TheManuz This is awesome! I'll start going through the changes, and will let you know if I spot anything that needs to be better. Standard code review procedure. :)

**Comment by TheManuz** [comment]
I'm sure the review is really needed, but i hope there won't be big problems! :D

File: src/input/pointer.js [hunk 1]
```diff
@@ -241,68 +237,141 @@
...
+											handled = true;
+											break;
+										}
+									}
+								}
+							}
+							break;
+					}
+				}
+			}
```
Comment (parasyte) [inline src/input/pointer.js]: Our coding conventions require 4 spaces instead of tab characters. Please convert all tabs to spaces in both files that were modified. That will help align the code properly on github, and make the rest of it easier to review.

Also, we have a file in the project root directory for editor configuration that will help for these kinds of settings: https://github.com/melonjs/melonJS/blob/master/.editorconfig

File: src/input/pointer.js [hunk 2]
```diff
@@ -314,8 +383,8 @@
     function updateCoordFromEvent(event) {
         var local;
 
-        // reset the touch array cache
-        changedTouches.length = 0;
+        // touch array cache
+        var changedTouches = [];
```
Comment (obiot) [inline src/input/pointer.js]: I would rather keep this one declared "globally" as before to avoid creating ana array object each time the function is called (this is why it was declared global before)

File: src/input/pointer.js [hunk 3]
```diff
@@ -560,7 +631,7 @@
      * // register on the 'pointerdown' event
      * me.input.registerPointerEvent('pointerdown', this, this.pointerDown.bind(this));
      */
-    obj.registerPointerEvent = function (eventType, rect, callback, floating) {
```
Comment (obiot) [inline src/input/pointer.js:563]: so if you removed the floating property, how does it work now ?

File: src/input/pointer.js [hunk 4]
```diff
@@ -572,25 +643,28 @@
...
+        if (!rect.GUID) {
+            rect.GUID = me.utils.createGUID();
+        }
+        
         // register the event
-        if (!evtHandlers[eventType]) {
-            evtHandlers[eventType] = [];
+        if (!evtHandlers[rect.GUID]) {
+            evtHandlers[rect.GUID] = {
+                rect : rect,
```
Comment (obiot) [inline src/input/pointer.js:659]: actually we should maybe change this part, change the function signature and replace the `rect` parameter by any object implementing the `getBounds()` function (and stores the object reference instead of the rect) and of course change any use of rect by a call to the object `getBounds()` function. 

Then If a `me.Rect` is provided, the behavior will probably be the same, but then if we pass an entity, it will allow to provide dynamic mouse detection for entities with changing shapes.

**Comment by obiot** [comment]
did you also test on iOS devices ? these are the tricky part with the standard (pointerEvent) emulation. Although older Android devices (and Cocoonjs) also use it so that should also work.

But globally the device/browser to test are :
* W7/8 - IE10 and IE11
* WIN - Chrome
* OSX - Chrome /Safari
* iOS 7/8 (webview and cocoonJS)
* Android (native browser and cocoonJS)

File: src/input/pointer.js (hunk 2 above)
Comment (TheManuz) [inline src/input/pointer.js]: You're right. In a previous implementation, when a move event needed to be re-dispatched as a enter/leave event, the global changedTouches caused many problems. I think I can switch it again to the global one, it should work with the current implementation.

File: src/input/pointer.js (hunk 3 above)
Comment (TheManuz) [inline src/input/pointer.js:563]: Since it saves a reference to the rect, it check if the rect is floating in the first place.

File: src/input/pointer.js (hunk 4 above)
Comment (TheManuz) [inline src/input/pointer.js:659]: I totally agree on this.

**Comment by TheManuz** [comment]
Ehm, sadly I'm working on an old laptop with Windows XP, so i cannot test on OSX. Also I've only got an Android smartphone, so i cannot test on iOS!
When I'll submit the new changes, I hope someone can test them on other platforms.

**Comment by parasyte** [comment]
No worries on compatibility testing. I think we can do that here before merging. :)

File: src/renderable/GUI.js [hunk 5]
```diff
@@ -146,16 +160,6 @@
...
-         * function callback for the pointerup event
-         * @ignore
-         */
-        release : function (event) {
-            this.released = true;
-            me.timer.clearTimeout(this.holdTimeout);
-            return this.onRelease(event);
-        },
-
-        /**
```
Comment (parasyte) [inline src/renderable/GUI.js:158]: nit: Why was this moved?

**Comment by obiot** [comment]
indeed the compatibility list was just a global reminder for all of us, based on my experience when working on this.

File: src/renderable/GUI.js (hunk 5 above)
Comment (TheManuz) [inline src/renderable/GUI.js:158]: Right question!
When i was working on the GUI_Object, i also added **moved**, **entered** and **leaved** function, and i moved all of them together with **clicked** and **release** to the top.
On the bottom i brought together the **onClick**, **onRelease** and the (now missing) **onMove**.
After some fiddling i removed unnecessary functions, but the order was left changed.

**Comment by obiot** [comment]
Hi @TheManuz, sorry i have not been able to follow on this one recently, so any update her, what is the status of this one, especially in terms of testing ?

File: src/input/pointer.js [hunk 6]
```diff
@@ -243,63 +245,136 @@
...
-            // get the current screen to world offset
-            me.game.viewport.localToWorld(0, 0, viewportOffset);
-            for (var t = 0, l = changedTouches.length; t < l; t++) {
-                // Do not fire older events
-                if (typeof(e.timeStamp) !== "undefined") {
-                    if (e.timeStamp < lastTimeStamp) {
-                        continue;
+        
+        for (var guid in evtHandlers) {
+            if (evtHandlers.hasOwnProperty(guid)) {
```
Comment (parasyte) [inline src/input/pointer.js:250]: This for-in-hasOwnProperty loop should be changed to:

``` javascript
Object.keys(evtHandlers).forEach(function (guid) {
    // ...
});
```

Comment (TheManuz) [inline src/input/pointer.js:250]: @parasyte: doesn't creating an anonymous function everytime adds overhead?
@obiot: testing has been done on Chrome and CocoonJS for Android.

**Comment by obiot** [comment]
@parasyte i'm hesitating here, do we merge this as part of 2.1, or do we keep it for 2.2 ?
]

**Comment by parasyte** [comment]
I haven't tested it, BUT I WANT IT.

**Comment by obiot** [comment]
ahah, i want it too :P so then in 2.1 ?

I'll let you do the honor of clicking on the merge button as your confirmation you want it now :)

After that we should enter into feature freeze for 2.1, else we will never release it :P

**Comment by parasyte** [comment]
@obiot @TheManuz merged!

I also fixed a few bugs along the way.

**Comment by obiot** [comment]
awesome ! thanks ;)

**Comment by TheManuz** [comment]
@parasyte thanks for doing the merge and fix bugs you encountered!
Sadly, i've got very little time lately, so i couldn't work on this! (I started the branch during christmas holidays, i've had much time then!).
Anyway, i'm happy! :smile:

**Comment by obiot** [comment]
@TheManuz thank you again for this great contribution :)
